acme/api/account.go

```diff
@@ -225,49 +237,47 @@ func (h *Handler) GetOrdersByAccountID(w http.ResponseWriter, r *http.Request) {
 	logOrdersByAccount(w, orders)
 }
 
-// validateExternalAccountBinding validates the externalAccountBinding property in a call to new-account
+// validateExternalAccountBinding validates the externalAccountBinding property in a call to new-account.
 func (h *Handler) validateExternalAccountBinding(ctx context.Context, nar *NewAccountRequest) (*acme.ExternalAccountKey, error) {
-
-	prov, err := provisionerFromContext(ctx)
+	acmeProv, err := acmeProvisionerFromContext(ctx)
 	if err != nil {
-		return nil, err
+		return nil, acme.WrapErrorISE(err, "could not load ACME provisioner from context")
 	}
 
-	acmeProv, ok := prov.(*provisioner.ACME) // TODO: rewrite into providing configuration via function on acme.Provisioner
-	if !ok || acmeProv == nil {
-		return nil, acme.NewErrorISE("provisioner in context is not an ACME provisioner")
-	}
-
-	shouldSkipAccountBindingValidation := !acmeProv.RequireEAB
-	if shouldSkipAccountBindingValidation {
+	if !acmeProv.RequireEAB {
 		return nil, nil
 	}
 
 	if nar.ExternalAccountBinding == nil {
 		return nil, acme.NewError(acme.ErrorExternalAccountRequiredType, "no external account binding provided")
 	}
 
+	// TODO: extract the EAB in a similar manner as JWS, JWK, payload, etc? That would probably move a lot/all of
+	// the logic of this function into the middleware. Should not be too hard, because the middleware does know
+	// about the handler and thus about its dependencies.
 	eabJSONBytes, err := json.Marshal(nar.ExternalAccountBinding)
 	if err != nil {
-		return nil, acme.WrapErrorISE(err, "error marshaling externalAccountBinding")
+		return nil, acme.WrapErrorISE(err, "error marshaling externalAccountBinding into JSON")
 	}
 
 	eabJWS, err := squarejose.ParseSigned(string(eabJSONBytes))
 	if err != nil {
 		return nil, acme.WrapErrorISE(err, "error parsing externalAccountBinding jws")
 	}
 
-	// TODO: verify supported algorithms against the incoming alg (and corresponding settings)?
 	// TODO: implement strategy pattern to allow for different ways of verification (i.e. webhook call) based on configuration
 
 	keyID := eabJWS.Signatures[0].Protected.KeyID
-	externalAccountKey, err := h.db.GetExternalAccountKey(ctx, keyID)
+	externalAccountKey, err := h.db.GetExternalAccountKey(ctx, acmeProv.Name, keyID)
 	if err != nil {
+		if _, ok := err.(*acme.Error); ok {
```

Comment:
Why do we need this extra error typing block? Maybe that will become obvious once I've seen db.GetExternalAccountKey.

Reply:
Within the lookup an acme.Error (`acme.ErrorUnauthorizedType`) is returned if the EAB credential does not belong to the provisioner. The additional type check allows an ACME client to interpret this specific error more clearly than without it. There may be a cleaner way to do this, but I haven't given that too much thought yet.